File: spring-boot-server/src/main/kotlin/com/witkups/carsharing/users/application/RoutePart.kt
package com.witkups.carsharing.users.application

import com.witkups.carsharing.users.user.ApplicationUser
import org.hibernate.annotations.Check
import javax.persistence.*

@Entity
@Table(name = "route_parts")
@Check(constraints = "departure <= arrival")
data class RoutePart(
  @Id
  @Column(name = "route_part_id")
  @GeneratedValue(strategy = GenerationType.AUTO)
  var id: Long? = null,

  @Column(nullable = false, name = "route_part_order")
  var order: Int? = null,

  @Embedded
  @AttributeOverrides(
    AttributeOverride(name = "date", column = Column(name = "departure", nullable = false)),
    AttributeOverride(name = "location_id", column = Column(name = "origin", nullable = false))
  )
  @AssociationOverride(
    name = "location", joinColumns = [
    JoinColumn(name = "origin_id", nullable = false,
//      insertable = false, updatable = false,
      foreignKey = ForeignKey(name = "FK_ROUTE_PARTS_ORIGIN_ID"))
  ])
  var origin: RouteSnapshot? = null,

  @Embedded
  @AttributeOverrides(
    AttributeOverride(name = "date", column = Column(name = "arrival", nullable = false)),
    AttributeOverride(name = "location_id", column = Column(name = "destination_id", nullable = false))
  )
  @AssociationOverride(
    name = "location", joinColumns = [
    JoinColumn(name = "destination_id", nullable = false,
//      insertable = false, updatable = false,
      foreignKey = ForeignKey(name = "FK_ROUTE_PARTS_DESTINATION_ID"))
  ])
  var destination: RouteSnapshot? = null,

  @Column(nullable = false)
  var distance: Double? = null,

  @Column(nullable = false)
  var cost: Double? = null,

  @ManyToMany(fetch = FetchType.LAZY)
  @JoinTable(name = "route_part_passengers",
    joinColumns = [(JoinColumn(name = "route_part_id", foreignKey = ForeignKey(name = "FK_APP_USER_ROUTE_PART")))],
    inverseJoinColumns = [(JoinColumn(name = "user_id", foreignKey = ForeignKey(name = "FK_ROUTE_PART_APP_USER")))]
  )
  var passengers: Set<ApplicationUser> = setOf()
)

File: frontend/src/app/functional/route/route-part.ts
import {RouteSnapshot} from "./route-snapshot";

export class RoutePart {
  constructor(
    public order: number,
    public distance: number,
    public cost: number,
    public origin: RouteSnapshot,
    public destination: RouteSnapshot
  ) {}

  static copy(r: RoutePart): RoutePart {
    return new RoutePart(
      r.order, r.distance, r.cost,
      RouteSnapshot.copy(r.origin),
      RouteSnapshot.copy(r.destination)
    )
  }

  getDuration() {
    if (this.isDateDefined('origin') && this.isDateDefined('destination')) {
      const dur = this.destination.date.getTime() - this.origin.date.getTime();
      if (dur > 0) {
        return dur;
      }
    }
    return null;
  }

  isDateDefined(field: string): boolean {
    return this[field] && this[field].date
  }

  isValid() {
    return this.isValidLocation('origin') && this.isValidLocation('destination')
  }

  isValidLocation(field:string) {
    return this[field] && this[field].location && this[field].location.label
  }
}
